homeassistant/components/media_player/template_media_player.py

```diff
@@ -0,0 +1,198 @@
+        self._volume_step = config.get(CONF_VOLUME_STEP)
+        self._volume_step_delay = config.get(CONF_VOLUME_STEP_DELAY)
+        self._is_muted = False
+
+        self._supported_features = 0
+        if CONF_ON_ACTION in self._actions:
+            self._supported_features |= SUPPORT_TURN_ON
+        if CONF_OFF_ACTION in self._actions:
+            self._supported_features |= SUPPORT_TURN_OFF
+        if (CONF_VOLUME_UP_ACTION in self._actions and
+            CONF_VOLUME_DOWN_ACTION in self._actions) or \
+            CONF_VOLUME_SET_ACTION in self._actions:
+            self._supported_features |= SUPPORT_VOLUME_SET
+        if CONF_VOLUME_MUTE_ACTION in self._actions and \
+           CONF_VOLUME_UNMUTE_ACTION in self._actions:
+            self._supported_features |= SUPPORT_VOLUME_MUTE
+        if self._sources and CONF_SELECT_SOURCE_ACTION in self._actions:
+            self._supported_features |= SUPPORT_SELECT_SOURCE
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return self._supported_features
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return list(self._sources.values())
+
+    def turn_on(self):
+        """Turn on media player."""
+        self._state = STATE_ON
+        if CONF_ON_ACTION in self._actions:
+            self._actions[CONF_ON_ACTION].async_run()
+
+    def turn_off(self):
+        """Turn off media player."""
+        self._state = STATE_OFF
+        if CONF_OFF_ACTION in self._actions:
+            self._actions[CONF_OFF_ACTION].async_run()
+
+    @property
+    def volume_level(self):
+        """Volume level of the media player (0..1)."""
+        return self._volume / 100.0
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._is_muted
+
+    def set_volume_level(self, volume):
+        volume = volume * 100
+        volume = int(math.ceil(volume / 10.0)) * 10
+        if CONF_VOLUME_SET_ACTION in self._actions:
+            self._volume = volume
+            self._actions[CONF_VOLUME_SET_ACTION].async_run(
+                {'volume:': volume})
+        elif CONF_VOLUME_UP_ACTION in self._actions and \
+                        CONF_VOLUME_UP_ACTION in self._actions:
+            adjustment_loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(adjustment_loop)
+            adjustment_loop.run_until_complete(asyncio.ensure_future(
+                self.adjust_volume(volume)
+            ))
+            adjustment_loop.close()
+
+    @asyncio.coroutine
+    def adjust_volume(self, volume):
+        if volume < self._volume:
+            while volume < self._volume:
```

Comment:
I don't think this volume stepping code belongs here at all. If the device doesn't support a volume_set command, we shouldn't try to fake it. Just expose volume up and down services.

Reply:
IIRC @pvizeli also had concerns with the volume stepping feature. This was one of the key reasons why I wrote this thing in the first place; I've been using this as a custom component to control two different amplifiers around the house. One (an Harman Kardon AVR1717) has an telnet interface, but no means of setting volumes to specific levels, only up or down. The other (a Scandyna The Amp) is controlled via IR and only supports volume up and down. Using this I can make it behave like other Media Players which is nice.
The alternative would be to add up and down buttons to the polymer component whenever a media player component has no `set_volume` operations, but does have `volume_up` and `volume_down`. Maybe that is a better solution?

Reply:
There should already be volume buttons on the more info card for volume stepping.

Reply:
@armills The volume control only appears if the media player supports the `SET_VOLUME` feature. That method receives the target volume as a numeric argument. The `MediaPlayerDevice` does not have support flags for volume stepping.

Reply:
https://github.com/home-assistant/home-assistant/blob/638dd25affb797037988adec94d7c0fbe4b90e92/homeassistant/components/media_player/__init__.py#L104

Reply:
@armills Perfect, thanks! How did I manage to miss that? The simulated volume stepping is now history.